cpp_ext/TorchDType.cpp

```diff
@@ -0,0 +1,168 @@
+//
+// Created by maksim on 6/13/23.
+//
+
+#include "IRModule.h"
+#include "TorchTypes.h"
+#include "mlir-c/BuiltinTypes.h"
+
+#include "TorchDType.h"
+
+#include <pybind11/cast.h>
+#include <pybind11/detail/common.h>
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+
+namespace py = pybind11;
+using namespace mlir::python;
+
+namespace mlir::torch {
+
+DType fromNpDType(const py::dtype &npDType) {
+  switch (npDType.num()) {
+  case 0: // numpy.bool_
+    return DType::bool_;
+  case 2: // numpy.uint8
+    return DType::uint8;
+  case 1: // numpy.int8
+    return DType::int8;
+  case 3: // numpy.int16
+    return DType::int16;
+  case 5: // numpy.int32
+    return DType::int32;
+  case 7: // numpy.int64
+    return DType::int64;
+  case 23: // numpy.float16
+    return DType::float16;
+  case 11: // numpy.float32
+    return DType::float32;
+  case 12: // numpy.float64
+    return DType::float64;
+  case 14: // numpy.complex64
+    return DType::complex32;
+  case 15: // numpy.complex128
+    return DType::complex64;
+  default:
+    throw py::value_error(py::repr(npDType).operator std::string() +
+                          " unsupported");
+  }
+}
+
+void populateTorchDType(py::module &m) {
+  py::enum_<DType>(m, "dtype", py::arithmetic())
+      .value("uint8", DType::uint8)
+      .value("int8", DType::int8)
+      .value("int16", DType::int16)
+      .value("int32", DType::int32)
+      .value("int64", DType::int64)
+      .value("float16", DType::float16)
+      .value("float32", DType::float32)
+      .value("float64", DType::float64)
+      .value("complex32", DType::complex32)
+      .value("complex64", DType::complex64)
+      .value("bool", DType::bool_)
+      .value("qint8", DType::qint8)
+      .value("quint8", DType::quint8)
+      .value("bfloat16", DType::bfloat16)
+      .def(
+          "to_mlir_type",
+          [](DType &self, DefaultingPyMlirContext context) {
+            switch (self) {
+            case bool_:
+              // default is signless
+              return mlirIntegerTypeGet(context->get(), 1);
+            case uint8:
+              return mlirIntegerTypeUnsignedGet(context->get(), 8);
+            case int8:
+              return mlirIntegerTypeSignedGet(context->get(), 8);
+            case int16:
+              return mlirIntegerTypeSignedGet(context->get(), 16);
+            case int32:
+              return mlirIntegerTypeSignedGet(context->get(), 32);
+            case int64:
+              return mlirIntegerTypeSignedGet(context->get(), 64);
```

Comment:
This is an important gotcha wrt torch-mlir: torch-mlir uses [signed integers](https://mlir.llvm.org/docs/Rationale/Rationale/#integer-signedness-semantics) at the type level but numpy has no such thing (the mapping here is basically the same as [`memref_type_to_np_dtype`](https://github.com/llvm/torch-mlir/blob/7c6961bcbf75811c21c026b31e2fe3eb67099cb9/python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py#L31C1-L40)